consensus/src/dag/dag_network.rs

```diff
@@ -28,5 +39,136 @@ pub trait DAGNetworkSender: Send + Sync {
         responders: Vec<Author>,
         message: ConsensusMsg,
         timeout: Duration,
-    ) -> anyhow::Result<ConsensusMsg>;
+    ) -> RpcWithFallback;
+}
+
+struct Responders {
+    peers: Vec<Author>,
+    generator: ExponentialNumberGenerator,
+}
+
+impl Responders {
+    fn new(mut peers: Vec<Author>, initial_request_count: u32, max_request_count: u32) -> Self {
+        peers.shuffle(&mut rand::thread_rng());
+        Self {
+            peers,
+            generator: ExponentialNumberGenerator::new(initial_request_count, 2, max_request_count),
+        }
+    }
+
+    fn next_to_request(&mut self) -> Option<Vec<Author>> {
+        let count = self.generator.next().expect("should return a number");
+
+        if self.peers.is_empty() {
+            return None;
+        }
+        Some(
+            self.peers.split_off(
+                self.peers
+                    .len()
+                    .checked_sub(count as usize)
+                    .unwrap_or(0),
+            ),
+        )
+    }
+}
+
+pub struct RpcWithFallback {
+    responders: Responders,
+    message: ConsensusMsg,
+    timeout: Duration,
+
+    terminated: bool,
+    futures: Pin<
+        Box<FuturesUnordered<Pin<Box<dyn Future<Output = anyhow::Result<ConsensusMsg>> + Send>>>>,
+    >,
+    sender: Arc<dyn DAGNetworkSender>,
+    interval: Pin<Box<Interval>>,
+}
+
+impl RpcWithFallback {
+    pub fn new(
+        responders: Vec<Author>,
+        message: ConsensusMsg,
+        timeout: Duration,
+        sender: Arc<dyn DAGNetworkSender>,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            responders: Responders::new(responders, 1, 4),
+            message,
+            timeout,
+
+            terminated: false,
+            futures: Box::pin(FuturesUnordered::new()),
+            sender,
+            interval: Box::pin(time_service.interval(timeout)),
+        }
+    }
+}
+
+impl Stream for RpcWithFallback {
+    type Item = anyhow::Result<ConsensusMsg>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let send_rpc = move |sender: Arc<dyn DAGNetworkSender>, peer, message, timeout| async move {
+            sender.send_rpc(peer, message, timeout).await
+        };
+
+        let timeout = matches!(self.interval.as_mut().poll_next(cx), Poll::Ready(_));
+
+        if self.futures.is_empty() || timeout {
```

Comment:
What if timeout expires but futures has more ready futures that were not pooled yet?

Reply:
are you saying I should poll the future first and then poll the timeout? sure

Reply:
BTW, why do you need this timeout? cannot we rely on the timeout inside the RPC?

Reply:
Good point/catch. I meant to split these into two timeouts. We should give a longer timeout for RPC to return and a shorter timeout for queuing more RPCs. The idea is in the best case we should expect a reply within the shorter timeout, but wait till the longer timeout to actually consider the RPC timed out. And, we will use the shorter timeout to queue more requests to improve the chances of getting a reply. I don't remember where but I read this is a way to reduce tail latencies. QS Batch requester does something similar.

Reply:
yes, we want two separate timeout for the request itself and sending next request